src/emmy/numerical/ode.cljc

```diff
@@ -2,96 +2,171 @@
+  (let [dimension (count y0)]
+    #?(:clj
+       (let [gbs               (GraggBulirschStoerIntegrator. 0. 1. (double epsilon) (double epsilon))
+             ode               (ExpandableStatefulODE.
+                                (reify FirstOrderDifferentialEquations
+                                  (computeDerivatives
+                                    [_ x y out]
+                                    (let [y' (f' x y)]
+                                      (doseq [i (range 0 (alength out))]
+                                        (aset out i (nth y' i)))))
+                                  (getDimension [_] dimension)))
+             step-requests     (a/chan)
+             solution-segments (a/chan)]
+         (doto ode
+           (.setTime x0)
+           (.setPrimaryState (double-array y0)))
+         (.addStepHandler
+          gbs
+          (reify StepHandler
+            (init [_ _ _ _])
+            (handleStep
+              [_ interpolator _]
+              ;; The `step-requests` channel sends `true` each time a new segment of 
+              ;; the solution is demanded; `false` is a signal that the consumer
+              ;; has no further need of them. When sending segments back, receiving
+              ;; false from the send operation indicates that the channel is closed.
+              ;; In either of these cases the integration should be shut down. 
+              ;;
+              ;; Normally, the use of an exception for flow control is frowned upon.
+              ;; Ideally, the CM3 integrator would look at the return value of the 
+              ;; step handler to decide whether or not to continue, but it doesn't.
+              ;; (You can install an EventHander, which watches an indicator function
+              ;; whose change of sign can halt the integration, but using that feature 
+              ;; for this purpose would be contrived, and the extra work the integrator 
+              ;; must do to watch the indicator function is not justified by this 
+              ;; goal.) Handling one exception in a worker thread is a small cost.
+              (when-not (and (a/>!! solution-segments
+                                    (let [x0 (.getPreviousTime interpolator)
+                                          x1 (.getCurrentTime interpolator)]
+                                      {:x0 x0
+                                       :x1 x1
+                                       :f  (fn [x]
+                                             (assert (and (>= x x0)
+                                                          (<= x x1))
+                                                     (format "ODE interpolation request %g out of range [%g, %g]"
+                                                             x x0 x1))
+                                             (.setInterpolatedTime interpolator x)
+                                             (into [] (.getInterpolatedState interpolator)))}))
```

Comment:
oh, yeah, I think I have the same feeling here, that `f'` should actually expect a double array, and should NOT keep a reference. Then we can wrap it to do this `(into [] ...)`.

The reason I want to do this is that I want to modify the `compile-fn` code (and did this in a branch already) to be able to generate `(aget y ...)` calls instead of destructuring or `nth` calls. This ends up being a good bit faster.

Reply:
I waffled on this. I can think of three different kinds of derivative.
- state derivatives (x is ignored, state is not flat)
- clojure (x matters, state is a seq of the customer's choosing, but probably `[...]`)
- native (x matters, state is a primitive array)
and my decision was that a good ODE library would be clojure-y. I suppose we can get away with it if we insist on having the user mutate the array given, so the input state f' receives is a primitive array but the user won't notice because it's a seq, same for the array that the user fills up. What I would lament slightly is not having the ability to write derivatives like `(x, [y0, y1]) => [-y1, y0]`

Reply:
done (and you might even get the compile-fn code to generate `(aset-double y ...)` on the way out